File: frontend/src/services/connections.ts
import { apiFetch } from "./api";

export interface ConnectionCreate {
  name: string;
  dbType: "postgres" | "mariadb";
  connectionUrl: string;
}

export interface ConnectionUpdate {
  name?: string;
  dbType?: "postgres" | "mariadb";
  connectionUrl?: string;
}

export interface ConnectionResponse {
  id: string;
  name: string;
  dbType: "postgres" | "mariadb";
  createdAt: string;
  lastUsedAt?: string | null;
  lastTestStatus: "success" | "failed" | "unknown";
}

export interface ConnectionListResponse {
  items: ConnectionResponse[];
  total: number;
  page: number;
  pageSize: number;
}

export interface ConnectionTestResponse {
  status: "success" | "failed";
  message?: string | null;
}

export function listConnections(): Promise<ConnectionListResponse> {
  return apiFetch("/connections");
}

export function createConnection(payload: ConnectionCreate): Promise<ConnectionResponse> {
  return apiFetch("/connections", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function deleteConnection(connectionId: string): Promise<void> {
  return apiFetch(`/connections/${connectionId}`, { method: "DELETE" });
}

export function updateConnection(
  connectionId: string,
  payload: ConnectionUpdate
): Promise<ConnectionResponse> {
  return apiFetch(`/connections/${connectionId}`, {
    method: "PUT",
    body: JSON.stringify(payload),
  });
}

export function testConnection(connectionId: string): Promise<ConnectionTestResponse> {
  return apiFetch(`/connections/${connectionId}/test`, { method: "POST" });
}

File: backend/src/models/connections.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field

from .base import AppBaseModel

DbType = Literal["postgres", "mariadb"]
TestStatus = Literal["success", "failed", "unknown"]


class ConnectionCreate(AppBaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    db_type: DbType
    connection_url: str = Field(..., min_length=1)


class ConnectionUpdate(AppBaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    db_type: DbType | None = None
    connection_url: str | None = Field(default=None, min_length=1)


class ConnectionResponse(AppBaseModel):
    id: str
    name: str
    db_type: DbType
    created_at: datetime
    last_used_at: datetime | None = None
    last_test_status: TestStatus


class ConnectionListResponse(AppBaseModel):
    items: list[ConnectionResponse]
    total: int
    page: int
    page_size: int


class ConnectionTestResponse(AppBaseModel):
    status: Literal["success", "failed"]
    message: str | None = None
